### PR TITLE
SequenceFileSink builder enhancements

### PR DESCRIPTION
**Pull request by davidak09**
I found some solution to the builder problem, so here's an enhancement proposal. It looks more readable, at least to me. :-)

Inspiration taken from: https://blog.jayway.com/2012/02/07/builder-pattern-with-a-twist/

### CODE REVIEW

**Comment by dmvk** [comment]
Interesting, it seems like a reasonable trade-off between code complexity and usability.

This could be probably also applied to generic type builders, eg.: ReduceByKey

cc @VaclavPlajt 

```
static <IN> KeyBuilder<IN>of(Dataset<IN> input) {
  // ...
}

interface KeyBuilder<IN> {
  <K> ReduceBuilder<IN, K> keyBy(UnaryFunction<IN, K> keyExtractor);
}

interface ReduceBuilder<IN, K> {
  <OUT> FinalBuilder<IN, K, OUT> keyBy(ReduceFunctor<IN, OUT> reducer);
}

interface FinalBuilder<IN, K, OUT> {
  // optinal builders
}

class Builder<IN, K, OUT> extends KeyBuilder<IN>, ReduceBuilder<IN, K>, FinalBuilder<IN, K, OUT> {
}
```

File: euphoria-hadoop/src/main/java/cz/seznam/euphoria/hadoop/output/SequenceFileSink.java [hunk 1]
```diff
@@ -52,263 +51,143 @@
...
-        Class<K> keyClass,
-        Class<V> valueType,
-        String outputPath,
-        boolean useLazyOutputFormat) {
-      this.keyClass = keyClass;
-      this.valueClass = valueType;
-      this.outputPath = outputPath;
-      this.useLazyOutputFormat = useLazyOutputFormat;
-    }
+  public interface Buildable<K, V> {
```
Comment (dmvk) [inline euphoria-hadoop/src/main/java/cz/seznam/euphoria/hadoop/output/SequenceFileSink.java]: Can we have a better name for this interface? For example `OptinalBuilder`? I also think that we should stick with `*Builder` naming convention for interfaces

Comment (davidak09) [reply euphoria-hadoop/src/main/java/cz/seznam/euphoria/hadoop/output/SequenceFileSink.java]: OK, I renamed `Buildable` to `OptionalBuilder` and `OutputPathBuildable` to `OutputPathBuilder`

**Comment by VaclavPlajt** [comment]
That is a clever solution to our builders problem. I elaborated little bit more. And it seem to suit to our needs. See [an example](https://github.com/VaclavPlajt/playground/blob/master/src/main/java/cz/plastique/examples/ClassToBuild.java).

**Comment by dmvk** [comment]
@VaclavPlajt awesome, this should simplify things a lot

**Pull request merged**